x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -281,6 +306,13 @@ public void onSkip() {
                     onResponse(new DataNodeComputeResponse(List.of(), Map.of()));
                 }
             }
+
+            private void maybeScheduleRetry(ShardId shardId, Exception e) {
+                if (targetShards.getShard(shardId).remainingNodes.isEmpty()
+                    && unwrapFailure(e) instanceof NoShardAvailableActionException) {
```

Comment:
This relies on `org.elasticsearch.xpack.esql.plugin.DataNodeRequestSender#unwrapFailure` that uses

https://github.com/elastic/elasticsearch/blob/a59c182f9f7e9d1bf3d6eecbc0e44f24ff91d053/server/src/main/java/org/elasticsearch/action/support/TransportActions.java#L22-L30

Probably this should only retry ShardNotFoundException/NoShardAvailableActionException/UnavailableShardsException. Not sure. Please let me know what do you think.